README.md
# string-to-svg

Rewrite of [text-to-svg](https://github.com/shrhdk/text-to-svg) in Typescript.

Convert text to SVG path without native dependence.

```ts
import { getSVG, loadFont } from 'string-to-svg'

const font = loadFont('path/to/font.ttf')
const attributes = { fill: 'red', stroke: 'black' };
const options = { font, x: 0, y: 0, fontSize: 72, anchor: 'top', attributes };

const svg = getSVG('hello', options);

console.log(svg);
```

```xml
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 180 72">
  <path fill="red" stroke="black" d="M5.27 9.28L10.62 9.28L10.62 29.36Q15.86 24.12 21.02 24.12Q26.89 24.12 29.60 29.29Q31.11 32.20 31.11 36.35L31.11 59.70L25.77 59.70L25.77 37.93Q25.77 29.21 20.18 29.21Q16.42 29.21 13.57 31.96Q10.62 34.91 10.62 38.71L10.62 59.70L5.27 59.70L5.27 9.28ZM67.68 49.08Q64.55 61.10 54.07 61.10Q47.57 61.10 43.77 55.69Q40.32 50.73 40.32 42.61Q40.32 34.84 43.56 29.88Q47.36 24.12 54 24.12Q66.97 24.12 67.82 43.70L45.74 43.70Q46.16 56.29 54.14 56.29Q60.47 56.29 62.05 49.08L67.68 49.08M62.05 39.09Q60.89 28.93 54 28.93Q47.36 28.93 45.95 39.09L62.05 39.09ZM92.81 51.82Q92.81 54.91 95.77 54.91Q98.19 54.91 101.07 54.35L101.07 59.73Q96.82 60.33 94.82 60.33Q87.19 60.33 87.19 52.84L87.19 9.28L92.81 9.28L92.81 51.82ZM128.81 51.82Q128.81 54.91 131.77 54.91Q134.19 54.91 137.07 54.35L137.07 59.73Q132.82 60.33 130.82 60.33Q123.19 60.33 123.19 52.84L123.19 9.28L128.81 9.28L128.81 51.82ZM162.07 24.12Q168.68 24.12 172.44 29.95Q175.68 34.80 175.68 42.61Q175.68 48.48 173.74 52.91Q170.16 61.14 161.93 61.14Q155.57 61.14 151.77 55.72Q148.32 50.77 148.32 42.61Q148.32 33.82 152.30 28.79Q156.09 24.12 162.07 24.12M161.93 29.14Q158.06 29.14 155.88 33.19Q153.95 36.74 153.95 42.61Q153.95 48.02 155.53 51.43Q157.71 56.11 162 56.11Q165.94 56.11 168.12 52.07Q170.05 48.52 170.05 42.68Q170.05 36.60 168.05 33.12Q165.90 29.14 161.93 29.14Z"/>
</svg>
```

## Installation

```
$ pnpm add --save string-to-svg
```

## API

### `getD(text, options)`

Get the path data for `d` attribute of `path`.

 - `text`: Text to convert to SVG path.

Options is an optional object containing:

 - `x`: Horizontal position of the beginning of the text. (default: `0`)
 - `y`: Vertical position of the baseline of the text. (default: `0`)
 - `fontSize`: Size of the text (default: `72`).
 - `kerning`: if `true` takes kerning information into account (default: `true`)
 - `letterSpacing`: letter-spacing value in `em`.
 - `tracking`: tracking value. (em / 1000)
 - `anchor`: Anchor of object in coordinate. (default: `'left baseline'`)
   - (`left`, `center`, `right`) + (`baseline`, `top`, `middle`, `bottom`)

Return value example

```
M5.27-54.07L10.62-54.07L10.62-34.00Q15.86-39.23 21.02-39.23Q26.89-39.23 29.60-34.07Q31.11-31.15 31.11-27L31.11-3.66L25.77-3.66L25.77-25.42Q25.77-34.14 20.18-34.14Q16.42-34.14 13.57-31.39Q10.62-28.44 10.62-24.64L10.62-3.66L5.27-3.66L5.27-54.07ZM67.68-14.27Q64.55-2.25 54.07-2.25Q47.57-2.25 43.77-7.66Q40.32-12.62 40.32-20.74Q40.32-28.51 43.56-33.47Q47.36-39.23 54-39.23Q66.97-39.23 67.82-19.65L45.74-19.65Q46.16-7.07 54.14-7.07Q60.47-7.07 62.05-14.27L67.68-14.27M62.05-24.26Q60.89-34.42 54-34.42Q47.36-34.42 45.95-24.26L62.05-24.26ZM92.81-11.53Q92.81-8.44 95.77-8.44Q98.19-8.44 101.07-9L101.07-3.62Q96.82-3.02 94.82-3.02Q87.19-3.02 87.19-10.51L87.19-54.07L92.81-54.07L92.81-11.53ZM128.81-11.53Q128.81-8.44 131.77-8.44Q134.19-8.44 137.07-9L137.07-3.62Q132.82-3.02 130.82-3.02Q123.19-3.02 123.19-10.51L123.19-54.07L128.81-54.07L128.81-11.53ZM162.07-39.23Q168.68-39.23 172.44-33.40Q175.68-28.55 175.68-20.74Q175.68-14.87 173.74-10.44Q170.16-2.21 161.93-2.21Q155.57-2.21 151.77-7.63Q148.32-12.59 148.32-20.74Q148.32-29.53 152.30-34.56Q156.09-39.23 162.07-39.23M161.93-34.21Q158.06-34.21 155.88-30.16Q153.95-26.61 153.95-20.74Q153.95-15.33 155.53-11.92Q157.71-7.24 162-7.24Q165.94-7.24 168.12-11.29Q170.05-14.84 170.05-20.67Q170.05-26.75 168.05-30.23Q165.90-34.21 161.93-34.21Z
```

### `getPath(text, options)`

Get the `path` element of SVG.

 - `text`: Text to convert to SVG path.

Options is an optional object containing:

 - `x`: Horizontal position of the beginning of the text. (default: `0`)
 - `y`: Vertical position of the baseline of the text. (default: `0`)
 - `fontSize`: Size of the text (default: `72`).
 - `kerning`: if `true` takes kerning information into account (default: `true`)
 - `letterSpacing`: letter-spacing value in `em`.
 - `tracking`: tracking value. (em / 1000)
 - `anchor`: Anchor of object in coordinate. (default: `'left baseline'`)
   - (`left`, `center`, `right`) + (`baseline`, `top`, `middle`, `bottom`)
 - `attributes`: Key-Value pairs of attributes of `path` element.

Return value example

```xml
<path fill="red" stroke="black" d="M5.27-54.07L10.62-54.07L10.62-34.00Q15.86-39.23 21.02-39.23Q26.89-39.23 29.60-34.07Q31.11-31.15 31.11-27L31.11-3.66L25.77-3.66L25.77-25.42Q25.77-34.14 20.18-34.14Q16.42-34.14 13.57-31.39Q10.62-28.44 10.62-24.64L10.62-3.66L5.27-3.66L5.27-54.07ZM67.68-14.27Q64.55-2.25 54.07-2.25Q47.57-2.25 43.77-7.66Q40.32-12.62 40.32-20.74Q40.32-28.51 43.56-33.47Q47.36-39.23 54-39.23Q66.97-39.23 67.82-19.65L45.74-19.65Q46.16-7.07 54.14-7.07Q60.47-7.07 62.05-14.27L67.68-14.27M62.05-24.26Q60.89-34.42 54-34.42Q47.36-34.42 45.95-24.26L62.05-24.26ZM92.81-11.53Q92.81-8.44 95.77-8.44Q98.19-8.44 101.07-9L101.07-3.62Q96.82-3.02 94.82-3.02Q87.19-3.02 87.19-10.51L87.19-54.07L92.81-54.07L92.81-11.53ZM128.81-11.53Q128.81-8.44 131.77-8.44Q134.19-8.44 137.07-9L137.07-3.62Q132.82-3.02 130.82-3.02Q123.19-3.02 123.19-10.51L123.19-54.07L128.81-54.07L128.81-11.53ZM162.07-39.23Q168.68-39.23 172.44-33.40Q175.68-28.55 175.68-20.74Q175.68-14.87 173.74-10.44Q170.16-2.21 161.93-2.21Q155.57-2.21 151.77-7.63Q148.32-12.59 148.32-20.74Q148.32-29.53 152.30-34.56Q156.09-39.23 162.07-39.23M161.93-34.21Q158.06-34.21 155.88-30.16Q153.95-26.61 153.95-20.74Q153.95-15.33 155.53-11.92Q157.71-7.24 162-7.24Q165.94-7.24 168.12-11.29Q170.05-14.84 170.05-20.67Q170.05-26.75 168.05-30.23Q165.90-34.21 161.93-34.21Z"/>
```

### `getSVG(text, options)`

Get the SVG.

 - `text`: Text to convert to SVG path.

Options is an optional object containing:

 - `x`: Horizontal position of the beginning of the text. (default: `0`)
 - `y`: Vertical position of the baseline of the text. (default: `0`)
 - `fontSize`: Size of the text (default: `72`).
 - `kerning`: if `true` takes kerning information into account (default: `true`)
 - `letterSpacing`: letter-spacing value in `em`.
 - `tracking`: tracking value. (em / 1000)
 - `anchor`: Anchor of object in coordinate. (default: `'left baseline'`)
   - (`left`, `center`, `right`) + (`baseline`, `top`, `middle`, `bottom`)
 - `attributes`: Key-Value pairs of attributes of `path` element.
 - `beforePath`: Content before the path.
 - `afterPath`: Content after the path.

Return value example

```xml
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><path fill="red" stroke="black" d="M5.27-54.07L10.62-54.07L10.62-34.00Q15.86-39.23 21.02-39.23Q26.89-39.23 29.60-34.07Q31.11-31.15 31.11-27L31.11-3.66L25.77-3.66L25.77-25.42Q25.77-34.14 20.18-34.14Q16.42-34.14 13.57-31.39Q10.62-28.44 10.62-24.64L10.62-3.66L5.27-3.66L5.27-54.07ZM67.68-14.27Q64.55-2.25 54.07-2.25Q47.57-2.25 43.77-7.66Q40.32-12.62 40.32-20.74Q40.32-28.51 43.56-33.47Q47.36-39.23 54-39.23Q66.97-39.23 67.82-19.65L45.74-19.65Q46.16-7.07 54.14-7.07Q60.47-7.07 62.05-14.27L67.68-14.27M62.05-24.26Q60.89-34.42 54-34.42Q47.36-34.42 45.95-24.26L62.05-24.26ZM92.81-11.53Q92.81-8.44 95.77-8.44Q98.19-8.44 101.07-9L101.07-3.62Q96.82-3.02 94.82-3.02Q87.19-3.02 87.19-10.51L87.19-54.07L92.81-54.07L92.81-11.53ZM128.81-11.53Q128.81-8.44 131.77-8.44Q134.19-8.44 137.07-9L137.07-3.62Q132.82-3.02 130.82-3.02Q123.19-3.02 123.19-10.51L123.19-54.07L128.81-54.07L128.81-11.53ZM162.07-39.23Q168.68-39.23 172.44-33.40Q175.68-28.55 175.68-20.74Q175.68-14.87 173.74-10.44Q170.16-2.21 161.93-2.21Q155.57-2.21 151.77-7.63Q148.32-12.59 148.32-20.74Q148.32-29.53 152.30-34.56Q156.09-39.23 162.07-39.23M161.93-34.21Q158.06-34.21 155.88-30.16Q153.95-26.61 153.95-20.74Q153.95-15.33 155.53-11.92Q157.71-7.24 162-7.24Q165.94-7.24 168.12-11.29Q170.05-14.84 170.05-20.67Q170.05-26.75 168.05-30.23Q165.90-34.21 161.93-34.21Z"/></svg>
```

### `getMetrics(text, option)`

Measure text size.

 - `text`: Text to measure size.

Options is an optional object containing:

 - `x`: Horizontal position of the beginning of the text. (default: `0`)
 - `y`: Vertical position of the baseline of the text. (default: `0`)
 - `fontSize`: Size of the text (default: `72`).
 - `kerning`: if `true` takes kerning information into account (default: `true`)
 - `letterSpacing`: letter-spacing value in `em`.
 - `tracking`: tracking value. (em / 1000)
 - `anchor`: Anchor of object in coordinate. (default: `'left baseline'`)

An example of return value.

```json
{
  "x": 0,
  "y": -63.3515625,
  "baseline": 0,
  "width": 180,
  "height": 72,
  "ascender": 63.3515625,
  "descender": -8.6484375
}
```

## License

MIT

## Credits

string-to-svg depends on the following softwares. I thank great authors a lot.

- [text-to-svg](https://github.com/shrhdk/text-to-svg): Copytight (c) 2016 Hideki Shiro
- [opentype.js](https://github.com/nodebox/opentype.js): Copyright (c) 2015 Frederik De Bleser

These are released under the [MIT license](https://opensource.org/licenses/MIT)


CHANGELOG.md
# CHANGELOG

## 4.0.1

 - Fix loadFont function.

## 4.0.0

**This version doesn't work. So do not use this version.**

 - Rewrite in Typescript by @sasnovacat08

## 3.1.5

 - Fix deployment script.

## ~~3.1.4~~

**This version had been failed to deploy. So do not use this version.**

 - Tested on Node.js v10.15.2
 - Update dependencies.
   - `opentype.js`: `0.7.3` -> `0.11.0`

## 3.1.3

 - Tested on Node.js v6.11.1
 - Update dependencies.
   - `opentype.js`: `0.7.1` -> `0.7.3`

## 3.1.2

 - Pin opentype.js version to `0.7.1`. (see #25)

## 3.1.1

 - Tested on Node.js v6.10.3
 - Update dependencies.
   - `opentype.js`: `0.6.6` -> `0.7.1`

## 3.1.0

 - Support letter-spacing option.
 - Support tracking option.

## 3.0.1

 - Fix #10

## 3.0.0

 - Simplify module layout
   - Please use `const TextToSVG = require('text-to-svg');`<br>instead of `const TextToSVG = require('text-to-svg').TextToSVG;`
 - Add `loadSync` method.
   - Please use `const textToSVG = TextToSVG.loadSync();`<br>instead of `const textToSVG = new TextToSVG();`
 - Add `load` method for asynchronous font loading.
 - Add `getMetrics` method instead of `getSize` method.
 - Correct vertical anchor option.
   - Correct `top` and `bottom`.
   - Add `baseline`.
 - Add `width` and `height` attribute to SVG element.

## 2.2.2

 - Tested on Node.js v5.3.0
 - Update dependencies.
   - `opentype.js`: `0.5.0` -> `0.6.0`

## 2.2.1

 - Tested on Node.js v4.2.1
 - Update dependencies.
 - Command line tool shows help when args are not provided.
 - Fix API document.

## 2.2.0

 - Add `getSize` method.
 - Add `anchor` option.
   - `left`, `center` , `right`, `top`, `middle`, `bottom`
 - Add command line tool.

## 2.1.0

 - Add `anchor` option.

## 2.0.2

 - Fix index.js

## 2.0.1

 - Include index.js to npm files.

## 2.0.0

 - Reform the API.
 - Include default font.
 - Remove the command line tool.

## 1.1.0

 - Add TextToSVG.getD(text, x, y, fontSize, options);

## 1.0.0

initial release


test/getSVG.test.ts
import { describe, it, expect } from 'vitest'
import { STSOptions, STSOptionsWithFont, getSVG as getSVGNode, loadFont as loadFontNode } from '../src/node'
import { getSVG as getSVGBrowser, loadFont as loadFontBrowser } from '../src/browser'
import { readFileSync } from 'fs'

describe('getSVGNode', () => {
  let word = ''
  function testGetSVG(options: STSOptions, expected: string) {
    const title = JSON.stringify(expected)
    const actual = getSVGNode(word, options)

    it(title, () => {
      expect(actual).toBe(expected)
    })
  }

  describe('hello', () => {
    word = 'hello'

    testGetSVG(
      { x: 0, y: 0, fontSize: 72, anchor: 'top', attributes: { fill: 'red', stroke: 'black' } },
      '<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="180" height="72"><path fill="red" stroke="black" d="M5.27 59.70L5.27 9.28L10.62 9.28L10.62 29.36Q15.86 24.12 21.02 24.12L21.02 24.12Q26.89 24.12 29.60 29.29L29.60 29.29Q31.11 32.20 31.11 36.35L31.11 36.35L31.11 59.70L25.77 59.70L25.77 37.93Q25.77 29.21 20.18 29.21L20.18 29.21Q16.42 29.21 13.57 31.96L13.57 31.96Q10.62 34.91 10.62 38.71L10.62 38.71L10.62 59.70L5.27 59.70ZM62.05 49.08L67.68 49.08Q64.55 61.10 54.07 61.10L54.07 61.10Q47.57 61.10 43.77 55.69L43.77 55.69Q40.32 50.73 40.32 42.61L40.32 42.61Q40.32 34.84 43.56 29.88L43.56 29.88Q47.36 24.12 54 24.12L54 24.12Q66.97 24.12 67.82 43.70L67.82 43.70L45.74 43.70Q46.16 56.29 54.14 56.29L54.14 56.29Q60.47 56.29 62.05 49.08L62.05 49.08ZM45.95 39.09L62.05 39.09Q60.89 28.93 54 28.93L54 28.93Q47.36 28.93 45.95 39.09L45.95 39.09ZM92.81 9.28L92.81 51.82Q92.81 54.91 95.77 54.91L95.77 54.91Q98.19 54.91 101.07 54.35L101.07 54.35L101.07 59.73Q96.82 60.33 94.82 60.33L94.82 60.33Q87.19 60.33 87.19 52.84L87.19 52.84L87.19 9.28L92.81 9.28ZM128.81 9.28L128.81 51.82Q128.81 54.91 131.77 54.91L131.77 54.91Q134.19 54.91 137.07 54.35L137.07 54.35L137.07 59.73Q132.82 60.33 130.82 60.33L130.82 60.33Q123.19 60.33 123.19 52.84L123.19 52.84L123.19 9.28L128.81 9.28ZM162.07 24.12L162.07 24.12Q168.68 24.12 172.44 29.95L172.44 29.95Q175.68 34.80 175.68 42.61L175.68 42.61Q175.68 48.48 173.74 52.91L173.74 52.91Q170.16 61.14 161.93 61.14L161.93 61.14Q155.57 61.14 151.77 55.72L151.77 55.72Q148.32 50.77 148.32 42.61L148.32 42.61Q148.32 33.82 152.30 28.79L152.30 28.79Q156.09 24.12 162.07 24.12ZM161.93 29.14L161.93 29.14Q158.06 29.14 155.88 33.19L155.88 33.19Q153.95 36.74 153.95 42.61L153.95 42.61Q153.95 48.02 155.53 51.43L155.53 51.43Q157.71 56.11 162 56.11L162 56.11Q165.94 56.11 168.12 52.07L168.12 52.07Q170.05 48.52 170.05 42.68L170.05 42.68Q170.05 36.60 168.05 33.12L168.05 33.12Q165.90 29.14 161.93 29.14Z"/></svg>',
    )

    const font = loadFontNode('../fonts/ipag.ttf')
    testGetSVG(
      { font, x: 0, y: 0, fontSize: 72, anchor: 'top', attributes: { fill: 'red', stroke: 'black' } },
      '<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="180" height="72"><path fill="red" stroke="black" d="M5.27 59.70L5.27 9.28L10.62 9.28L10.62 29.36Q15.86 24.12 21.02 24.12L21.02 24.12Q26.89 24.12 29.60 29.29L29.60 29.29Q31.11 32.20 31.11 36.35L31.11 36.35L31.11 59.70L25.77 59.70L25.77 37.93Q25.77 29.21 20.18 29.21L20.18 29.21Q16.42 29.21 13.57 31.96L13.57 31.96Q10.62 34.91 10.62 38.71L10.62 38.71L10.62 59.70L5.27 59.70ZM62.05 49.08L67.68 49.08Q64.55 61.10 54.07 61.10L54.07 61.10Q47.57 61.10 43.77 55.69L43.77 55.69Q40.32 50.73 40.32 42.61L40.32 42.61Q40.32 34.84 43.56 29.88L43.56 29.88Q47.36 24.12 54 24.12L54 24.12Q66.97 24.12 67.82 43.70L67.82 43.70L45.74 43.70Q46.16 56.29 54.14 56.29L54.14 56.29Q60.47 56.29 62.05 49.08L62.05 49.08ZM45.95 39.09L62.05 39.09Q60.89 28.93 54 28.93L54 28.93Q47.36 28.93 45.95 39.09L45.95 39.09ZM92.81 9.28L92.81 51.82Q92.81 54.91 95.77 54.91L95.77 54.91Q98.19 54.91 101.07 54.35L101.07 54.35L101.07 59.73Q96.82 60.33 94.82 60.33L94.82 60.33Q87.19 60.33 87.19 52.84L87.19 52.84L87.19 9.28L92.81 9.28ZM128.81 9.28L128.81 51.82Q128.81 54.91 131.77 54.91L131.77 54.91Q134.19 54.91 137.07 54.35L137.07 54.35L137.07 59.73Q132.82 60.33 130.82 60.33L130.82 60.33Q123.19 60.33 123.19 52.84L123.19 52.84L123.19 9.28L128.81 9.28ZM162.07 24.12L162.07 24.12Q168.68 24.12 172.44 29.95L172.44 29.95Q175.68 34.80 175.68 42.61L175.68 42.61Q175.68 48.48 173.74 52.91L173.74 52.91Q170.16 61.14 161.93 61.14L161.93 61.14Q155.57 61.14 151.77 55.72L151.77 55.72Q148.32 50.77 148.32 42.61L148.32 42.61Q148.32 33.82 152.30 28.79L152.30 28.79Q156.09 24.12 162.07 24.12ZM161.93 29.14L161.93 29.14Q158.06 29.14 155.88 33.19L155.88 33.19Q153.95 36.74 153.95 42.61L153.95 42.61Q153.95 48.02 155.53 51.43L155.53 51.43Q157.71 56.11 162 56.11L162 56.11Q165.94 56.11 168.12 52.07L168.12 52.07Q170.05 48.52 170.05 42.68L170.05 42.68Q170.05 36.60 168.05 33.12L168.05 33.12Q165.90 29.14 161.93 29.14Z"/></svg>',
    )
  })
})

describe('getSVGBrowser', () => {
  let word = ''
  function testGetSVG(options: STSOptionsWithFont, expected: string) {
    const title = JSON.stringify(expected)
    const actual = getSVGBrowser(word, options)

    it(title, () => {
      expect(actual).toBe(expected)
    })
  }

  describe('hello', () => {
    word = 'hello'

    const file = readFileSync('./fonts/ipag.ttf')
    const font = loadFontBrowser(file.buffer)
    testGetSVG(
      { font, x: 0, y: 0, fontSize: 72, anchor: 'top', attributes: { fill: 'red', stroke: 'black' } },
      '<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="180" height="72"><path fill="red" stroke="black" d="M5.27 59.70L5.27 9.28L10.62 9.28L10.62 29.36Q15.86 24.12 21.02 24.12L21.02 24.12Q26.89 24.12 29.60 29.29L29.60 29.29Q31.11 32.20 31.11 36.35L31.11 36.35L31.11 59.70L25.77 59.70L25.77 37.93Q25.77 29.21 20.18 29.21L20.18 29.21Q16.42 29.21 13.57 31.96L13.57 31.96Q10.62 34.91 10.62 38.71L10.62 38.71L10.62 59.70L5.27 59.70ZM62.05 49.08L67.68 49.08Q64.55 61.10 54.07 61.10L54.07 61.10Q47.57 61.10 43.77 55.69L43.77 55.69Q40.32 50.73 40.32 42.61L40.32 42.61Q40.32 34.84 43.56 29.88L43.56 29.88Q47.36 24.12 54 24.12L54 24.12Q66.97 24.12 67.82 43.70L67.82 43.70L45.74 43.70Q46.16 56.29 54.14 56.29L54.14 56.29Q60.47 56.29 62.05 49.08L62.05 49.08ZM45.95 39.09L62.05 39.09Q60.89 28.93 54 28.93L54 28.93Q47.36 28.93 45.95 39.09L45.95 39.09ZM92.81 9.28L92.81 51.82Q92.81 54.91 95.77 54.91L95.77 54.91Q98.19 54.91 101.07 54.35L101.07 54.35L101.07 59.73Q96.82 60.33 94.82 60.33L94.82 60.33Q87.19 60.33 87.19 52.84L87.19 52.84L87.19 9.28L92.81 9.28ZM128.81 9.28L128.81 51.82Q128.81 54.91 131.77 54.91L131.77 54.91Q134.19 54.91 137.07 54.35L137.07 54.35L137.07 59.73Q132.82 60.33 130.82 60.33L130.82 60.33Q123.19 60.33 123.19 52.84L123.19 52.84L123.19 9.28L128.81 9.28ZM162.07 24.12L162.07 24.12Q168.68 24.12 172.44 29.95L172.44 29.95Q175.68 34.80 175.68 42.61L175.68 42.61Q175.68 48.48 173.74 52.91L173.74 52.91Q170.16 61.14 161.93 61.14L161.93 61.14Q155.57 61.14 151.77 55.72L151.77 55.72Q148.32 50.77 148.32 42.61L148.32 42.61Q148.32 33.82 152.30 28.79L152.30 28.79Q156.09 24.12 162.07 24.12ZM161.93 29.14L161.93 29.14Q158.06 29.14 155.88 33.19L155.88 33.19Q153.95 36.74 153.95 42.61L153.95 42.61Q153.95 48.02 155.53 51.43L155.53 51.43Q157.71 56.11 162 56.11L162 56.11Q165.94 56.11 168.12 52.07L168.12 52.07Q170.05 48.52 170.05 42.68L170.05 42.68Q170.05 36.60 168.05 33.12L168.05 33.12Q165.90 29.14 161.93 29.14Z"/></svg>',
    )
  })
})


test/getPath.test.ts
import { describe, it, expect } from 'vitest'
import { STSOptionsWithFont, getPath } from '../src'
import { loadSync } from 'opentype.js'

const font = loadSync('./fonts/ipag.ttf')

describe('getPath', () => {
  let word = ''
  function testGetPath(options: STSOptionsWithFont, expected: string) {
    const title = JSON.stringify(expected)
    const actual = getPath(word, options)

    it(title, () => {
      expect(actual).toBe(expected)
    })
  }

  describe('hello', () => {
    word = 'hello'

    testGetPath(
      { font, x: 0, y: 0, fontSize: 72, anchor: 'top', attributes: { fill: 'red', stroke: 'black' } },
      '<path fill="red" stroke="black" d="M5.27 59.70L5.27 9.28L10.62 9.28L10.62 29.36Q15.86 24.12 21.02 24.12L21.02 24.12Q26.89 24.12 29.60 29.29L29.60 29.29Q31.11 32.20 31.11 36.35L31.11 36.35L31.11 59.70L25.77 59.70L25.77 37.93Q25.77 29.21 20.18 29.21L20.18 29.21Q16.42 29.21 13.57 31.96L13.57 31.96Q10.62 34.91 10.62 38.71L10.62 38.71L10.62 59.70L5.27 59.70ZM62.05 49.08L67.68 49.08Q64.55 61.10 54.07 61.10L54.07 61.10Q47.57 61.10 43.77 55.69L43.77 55.69Q40.32 50.73 40.32 42.61L40.32 42.61Q40.32 34.84 43.56 29.88L43.56 29.88Q47.36 24.12 54 24.12L54 24.12Q66.97 24.12 67.82 43.70L67.82 43.70L45.74 43.70Q46.16 56.29 54.14 56.29L54.14 56.29Q60.47 56.29 62.05 49.08L62.05 49.08ZM45.95 39.09L62.05 39.09Q60.89 28.93 54 28.93L54 28.93Q47.36 28.93 45.95 39.09L45.95 39.09ZM92.81 9.28L92.81 51.82Q92.81 54.91 95.77 54.91L95.77 54.91Q98.19 54.91 101.07 54.35L101.07 54.35L101.07 59.73Q96.82 60.33 94.82 60.33L94.82 60.33Q87.19 60.33 87.19 52.84L87.19 52.84L87.19 9.28L92.81 9.28ZM128.81 9.28L128.81 51.82Q128.81 54.91 131.77 54.91L131.77 54.91Q134.19 54.91 137.07 54.35L137.07 54.35L137.07 59.73Q132.82 60.33 130.82 60.33L130.82 60.33Q123.19 60.33 123.19 52.84L123.19 52.84L123.19 9.28L128.81 9.28ZM162.07 24.12L162.07 24.12Q168.68 24.12 172.44 29.95L172.44 29.95Q175.68 34.80 175.68 42.61L175.68 42.61Q175.68 48.48 173.74 52.91L173.74 52.91Q170.16 61.14 161.93 61.14L161.93 61.14Q155.57 61.14 151.77 55.72L151.77 55.72Q148.32 50.77 148.32 42.61L148.32 42.61Q148.32 33.82 152.30 28.79L152.30 28.79Q156.09 24.12 162.07 24.12ZM161.93 29.14L161.93 29.14Q158.06 29.14 155.88 33.19L155.88 33.19Q153.95 36.74 153.95 42.61L153.95 42.61Q153.95 48.02 155.53 51.43L155.53 51.43Q157.71 56.11 162 56.11L162 56.11Q165.94 56.11 168.12 52.07L168.12 52.07Q170.05 48.52 170.05 42.68L170.05 42.68Q170.05 36.60 168.05 33.12L168.05 33.12Q165.90 29.14 161.93 29.14Z"/>',
    )
  })
})


src/node.ts
import { join } from 'path'
import { loadSync, Font } from 'opentype.js'
import { Anchor, STSOptions, STSOptionsWithFont, getD, getHeight, getMetrics, getPath, getWidth } from './index'

export { Anchor, STSOptions, STSOptionsWithFont, getD, getHeight, getMetrics, getPath, getWidth }
export { Font }

export function loadFont(fontPath: string) {
  const path = join(__dirname, fontPath)
  return loadSync(path)
}

export function getSVG(text: string, options: STSOptions) {
  if (!options.font) {
    options.font = loadFont('../fonts/ipag.ttf')
  }
  const metrics = getMetrics(text, options as STSOptionsWithFont)
  let svg = `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="${metrics.width}" height="${metrics.height}">`
  if (options.beforePath) svg += options.beforePath
  svg += getPath(text, options as STSOptionsWithFont)
  if (options.afterPath) svg += options.afterPath
  svg += '</svg>'

  return svg
}


src/index.ts
import { Font } from 'opentype.js'

export type Anchor =
  | 'left'
  | 'center'
  | 'right'
  | 'top'
  | 'middle'
  | 'bottom'
  | 'left baseline'
  | 'left top'
  | 'left middle'
  | 'left bottom'
  | 'center top'
  | 'center middle'
  | 'center bottom'
  | 'right top'
  | 'right middle'
  | 'right bottom'

export interface STSOptions {
  font?: Font
  x?: number
  y?: number
  fontSize?: number
  anchor?: Anchor
  kerning?: boolean
  letterSpacing?: number
  tracking?: number
  attributes?: Record<string, string>
  beforePath?: string
  afterPath?: string
}

export type STSOptionsWithFont = STSOptions & { font: Font }

export function getD(text: string, options: STSOptionsWithFont) {
  const fontSize = options.fontSize || 72
  const kerning = options.kerning
  const letterSpacing = options.letterSpacing
  const tracking = options.tracking
  const metrics = getMetrics(text, options)
  const path = options.font.getPath(text, metrics.x, metrics.baseline, fontSize, { kerning, letterSpacing, tracking })

  return path.toPathData(2)
}

export function getPath(text: string, options: STSOptionsWithFont) {
  const attributes = options.attributes || {}
  const attributesString = Object.keys(attributes)
    .map((key) => `${key}="${attributes[key]}"`)
    .join(' ')
  const d = getD(text, options)

  if (attributesString) {
    return `<path ${attributesString} d="${d}"/>`
  }

  return `<path d="${d}"/>`
}

export function getHeight(fontSize: number, font: Font) {
  const fontScale = (1 / font.unitsPerEm) * fontSize
  return (font.ascender - font.descender) * fontScale
}

export function getWidth(text: string, options: STSOptionsWithFont) {
  const fontSize = options.fontSize || 72
  const kerning = 'kerning' in options ? options.kerning : true
  const fontScale = (1 / options.font.unitsPerEm) * fontSize

  let width = 0
  const glyphs = options.font.stringToGlyphs(text)
  for (let i = 0; i < glyphs.length; i++) {
    const glyph = glyphs[i]

    if (glyph.advanceWidth) {
      width += glyph.advanceWidth * fontScale
    }

    if (kerning && i < glyphs.length - 1) {
      const kerningValue = options.font.getKerningValue(glyph, glyphs[i + 1])
      width += kerningValue * fontScale
    }

    if (options.letterSpacing) {
      width += options.letterSpacing * fontSize
    } else if (options.tracking) {
      width += (options.tracking / 1000) * fontSize
    }
  }
  return width
}

function parseAnchorOption(anchor: Anchor) {
  let horizontalMatch = anchor.match(/left|center|right/gi) || []
  let horizontal = horizontalMatch.length === 0 ? 'left' : horizontalMatch[0]

  let verticalMatch = anchor.match(/baseline|top|bottom|middle/gi) || []
  let vertical = verticalMatch.length === 0 ? 'baseline' : verticalMatch[0]

  return { horizontal, vertical }
}

export function getMetrics(text: string, options: STSOptionsWithFont) {
  const fontSize = options.fontSize || 72
  const anchor = parseAnchorOption(options.anchor || 'left baseline')

  const width = getWidth(text, options)
  const height = getHeight(fontSize, options.font)

  const fontScale = (1 / options.font.unitsPerEm) * fontSize
  const ascender = options.font.ascender * fontScale
  const descender = options.font.descender * fontScale

  let x = options.x || 0
  switch (anchor.horizontal) {
    case 'left':
      x -= 0
      break
    case 'center':
      x -= width / 2
      break
    case 'right':
      x -= width
      break
    default:
      throw new Error(`Unknown anchor option: ${anchor.horizontal}`)
  }

  let y = options.y || 0
  switch (anchor.vertical) {
    case 'baseline':
      y -= ascender
      break
    case 'top':
      y -= 0
      break
    case 'middle':
      y -= height / 2
      break
    case 'bottom':
      y -= height
      break
    default:
      throw new Error(`Unknown anchor option: ${anchor.vertical}`)
  }

  const baseline = y + ascender

  return {
    x,
    y,
    baseline,
    width,
    height,
    ascender,
    descender,
  }
}
